main_test.go

```diff
@@ -0,0 +1,262 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package main
+
+import (
+	"io/ioutil"
+	"os"
+	"testing"
+
+	firecracker "github.com/firecracker-microvm/firecracker-go-sdk"
+	models "github.com/firecracker-microvm/firecracker-go-sdk/client/models"
+)
+
+func TestParseBlockDevices(t *testing.T) {
+	tempFile, err := ioutil.TempFile("", "firectl-test-drive-path")
+	if err != nil {
+		t.Error(err)
+	}
+	defer os.Remove(tempFile.Name())
+	defer tempFile.Close()
+	validDrive := models.Drive{
+		DriveID:      firecracker.String("2"),
+		PathOnHost:   firecracker.String(tempFile.Name()),
+		IsReadOnly:   firecracker.Bool(false),
+		IsRootDevice: firecracker.Bool(false),
+	}
+	cases := []struct {
+		in          []string
+		outDrives   []models.Drive
+		expectedErr func(error) bool
+	}{
+		// no suffix
+		{
+			in:        []string{"/path"},
+			outDrives: nil,
+			expectedErr: func(a error) bool {
+				return a == invalidDriveSpecificationNoSuffix
+			},
+		},
+		// no path
+		{
+			in:        []string{":rw"},
+			outDrives: nil,
+			expectedErr: func(a error) bool {
+				return a == invalidDriveSpecificationNoPath
+			},
+		},
+		// path does not exist
+		{
+			in:          []string{"/does/not/exist:ro"},
+			outDrives:   nil,
+			expectedErr: os.IsNotExist,
+		},
+		// valid
+		{
+			in:        []string{tempFile.Name() + ":rw"},
+			outDrives: []models.Drive{validDrive},
+			expectedErr: func(a error) bool {
+				return a == nil
+			},
+		},
+	}
+	for _, c := range cases {
+		drives, err := parseBlockDevices(c.in)
+		if !equalDrives(c.outDrives, drives) {
+			t.Errorf("expected %v but got %v for %s", c.outDrives, drives, c.in)
+		}
+		if !c.expectedErr(err) {
+			t.Errorf("did not get the expected err but received %s for %s", err, c.in)
+		}
+	}
+}
+
+func equalDrives(a, b []models.Drive) bool {
+	if a == nil && b == nil {
+		return true
+	}
+	if (a == nil && b != nil) || (a != nil && b == nil) {
+		return false
+	}
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if !equalDrive(a[i], b[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+func equalDrive(a, b models.Drive) bool {
+	if !equalString(a.DriveID, b.DriveID) ||
+		!equalString(a.PathOnHost, b.PathOnHost) ||
+		!equalBool(a.IsReadOnly, b.IsReadOnly) ||
+		!equalBool(a.IsRootDevice, b.IsRootDevice) {
+		return false
+	}
+	return true
+}
+
+func equalString(a, b *string) bool {
+	if a == nil && b == nil {
+		return true
+	}
+	if (a == nil && b != nil) || (a != nil && b == nil) {
```

Comment:
having `if a == nil && b == nil` we can write `else if a == nil || b == nil {`

Reply:
Code was removed in refactor.